src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs

```diff
@@ -15,27 +16,33 @@ public class CompilationWithAnalyzers
         private readonly Compilation _compilation;
         private readonly CancellationToken _cancellationToken;
         private readonly ConcurrentSet<Diagnostic> _exceptionDiagnostics;
+        private ImmutableHashSet<SyntaxTree> _analyzedSyntaxTrees = ImmutableHashSet<SyntaxTree>.Empty;
```

Comment:
Probably this can be a regular HashSet now that all accesses to it are guarded by locks?